### PR TITLE
Speed up most commonly used deriv/deriv2 in PGI

### PR DESCRIPTION
**Pull request by johnweis0480**
#### Summary
Avoid slow for loops over numpy arrays in PGI derivs. The associated changes in the __call__ can also be sped up but I left them as is for now to show the deriv tests work.

#### PR Checklist
* [x] If this is a work in progress PR, set as a Draft PR
* [x] Linted my code according to the [style guides](https://docs.simpeg.xyz/latest/content/getting_started/contributing/code-style.html).
* [x] Added [tests](https://docs.simpeg.xyz/latest/content/getting_started/contributing/testing.html) to verify changes to the code.
* [x] Added necessary documentation to any new functions/classes following the
      expect [style](https://docs.simpeg.xyz/latest/content/getting_started/contributing/documentation.html).
* [x] Marked as ready for review (if this is was a draft PR), and converted 
      to a Pull Request
* [x] Tagged ``@simpeg/simpeg-developers`` when ready for review.

### Squash and Merge summary

Speed up dot products in the derivative methods of the PGI regularization classes by making use of Numpy functions instead of for loops that perform a significant number of iterations.

### CODE REVIEW

File: simpeg/regularization/pgi.py [hunk 1]
```diff
@@ -606,14 +597,12 @@ def deriv(self, m):
...
-                                np.dot(self.gmm.precisions_[membership[i]], r0[i])
-                                for i in range(len(r0))
-                            ]
-                        ]
-                    )
+                    r = np.zeros_like(r0)
+                    for i in range(self.gmm.n_components):
+                        r[membership == i] = (
+                            self.gmm.precisions_[i] @ r0[membership == i].T
+                        ).T
```
Comment (santisoler) [inline simpeg/regularization/pgi.py:604]: Similar improvement here:
```suggestion
                        selection = membership == i
                        r[selection] = (
                            self.gmm.precisions_[i] @ r0[selection].T
                        ).T
```

**Review by santisoler — APPROVED** [review]
Nice catches @johnweis0480! Great job here.

I left only two minor improvements that aren't critical. Let me know what do you think.

Do you plan to add similar changes to the `__call__` method in a separate PR or would you like to do so here? Anyway, I think this PR is ready to go.

I think @thibaut-kobold might want to take a look at them as well.

**Comment by thibaut-kobold** [comment]
Thanks @johnweis0480 !

**Comment by johnweis0480** [comment]
@santisoler I can do them in another PR. Hey @thibaut-kobold, I also noticed in that the doc string the approx_gradient and approx_eval seem to be wrong. Can you double check this?

**Comment by thibaut-kobold** [comment]
> I also noticed in that the doc string the approx_gradient and approx_eval seem to be wrong. Can you double check this?

I won't be able to have a deep look today. Don't block the PR for it. Also, feel free to make a PR to fix them to what you think is right, and I will review that (probably the best way to tdo)

**Comment by santisoler** [comment]
Merged 🚀 Congrats @johnweis0480!

**Pull request merged**